### PR TITLE
HPCC-24250 TechPreview: Simplify ESP running as an application.

### PR DESCRIPTION
**Pull request by afishbeck**
Simplify running ESP as an application (such as EclWatch).

Currently will default ldap and tls both on.
Don't want users accidentally running insecure instances
when they don't mean to.
```
  – to turn off tls use the option "–tls=off"
  – to turn off ldap use the option "–auth=none"

```
First set of named applications will be:
```
  eclwatch
       the user facing EclWatch application.
  eclservices
       the same services that make up eclwatch, but with no user interface.
       to be used by HPCC components needing to make API calls
  eclqueries
       UI for running published ECL queries (wsecl)

```
To actually use tls you need certificates either in the default location, or by overriding the config.
Defaults:

```
 tls_config:
  certificate: /etc/HPCCSystems/certificates/{$name}/server.crt
  privatekey: /etc/HPCCSystems/certificates/{$name}/private.key
  passphrase: JHjdi6DDptMwYmJNTZsqjg==

```
NOTE: Need to figure out how to best provide passphrase..
..including looking at using Kubernetes TLS secrets which contain
all of this info together.

To use ldap you at least need to set the ldap server:
`--ldapAddress=<ldap_server>
`
Examples:
No ldap, no tls (both default on):
`    esp --port=8888 --application=eclwatch --daliServers=. --auth=none --tls=off
`
Ldap and tls.. explict port (--tlsPort would have worked as well)
`    esp --port=8888 --application=eclservices --daliServers=. --ldapAddress=ldap
`
No ldap, yes tls, pass cert/key/passphrase via command line.  (TBD improve passphrase handling)
```
    esp --port=8888 --application=eclwatch --daliServers=. --auth=none
        --tls_config.certificate=/home/tony/certificates/server.crt
        --tls_config.privatekey=/home/tony/certificates/server.key
        --tls_config.passphrase=JHjdi6DDptMwYmJNTZsqjg==

```
You can also use your own app config yaml file containing only those items you want to override from the default.  For example:

myesp.yaml containing:
```
esp:
  port: 8888
  daliServers: .
  auth: none
  tls_config:
    certificate: /home/tony/certificates/server.crt
    privatekey: /home/tony/certificates/server.key
    passphrase: JHjdi6DDptMwYmJNTZsqjg==

```
---------------------------------------------
To run using myesp.yaml:
`    esp --application=eclqueries --config=myesp.yaml
`

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by afishbeck** [comment]
@richardkchapman especially looking for design and usability feedback at this point.

**Comment by afishbeck** [comment]
I've left the task of cleaning up service specific config items for service owners to do later.

**Comment by afishbeck** [comment]
Rebased to latest master and fixed esp.scm issue.

**Comment by cloLN** [comment]
Just making sure the encrypted passwords in the ldap.yaml files aren't actual passwords in use

**Comment by cloLN** [comment]
has the issue with running preflight causing a stack trace been resolved?  It actually kills the eclwatch application that's running

File: esp/applications/eclwatch/eclwatch.yaml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      ActivityInfoCacheAutoRebuildSeconds: 120
+   WsWorkunits:
+      ViewTimeout: 1000
+      LayoutProgram: dot/dot -Tsvg -Gordering=out
+      EclServer: ''
+      EclServerQueue: ''
+      StyleSheets:
+        xslt:
+        - name: atts
+          ^: "/esp/xslt/atts.xslt"
```
Comment (richardkchapman) [inline esp/applications/eclwatch/eclwatch.yaml:13]: I'm not familiar with this yaml syntax. What does it mean?

File: esp/applications/eclwatch/ldap_authorization_map.yaml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            resource: DfuExceptions
+            description: Access to DFU exceptions
+         -  path: FileSprayAccess
+            resource: FileSprayAccess
+            description: Access to spraying files
+         -  path: FileDesprayAccess
+            resource: FileDesprayAccess
+            description: Access to despraying of files
+         -  path: FileDkcAccess
+            resource: FileDkcAccess
```
Comment (richardkchapman) [inline esp/applications/eclwatch/ldap_authorization_map.yaml:122]: Do we still support DKC ??

File: esp/applications/hpcc_services/hpcc_services.yaml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            description: Generic KeyVal store for HPCC Applications
+            default: true
+   ws_machine:
+      excludePartitions: /dev*,/sys*,/proc*,/run*,/boot
+      monitorDaliFileServer: false
+      UseDefaultSSHUserID: false
+      UseDefaultPIDFileName: true
+      ProcessFilters:
+        Platform:
+        - name: Windows
```
Comment (richardkchapman) [inline esp/applications/hpcc_services/hpcc_services.yaml:71]: Windows? 

What does this section mean anyway?

**Comment by richardkchapman** [comment]
Do we need to make some of this conditional on being in containerized mode, or are we still fully compatible with legacy installations using configmgr etc

File: esp/services/WsDeploy/WsDeployPlugin.cpp [hunk 4]
```diff
@@ -44,7 +44,7 @@ extern "C"
 // Change the function names when we stick with dynamic loading.
 ESP_FACTORY IEspService * esp_service_factory(const char *name, const char* type, IPropertyTree *cfg, const char *process)
 {
-   if (strcmp(type, "WsDeploy")==0)
```
Comment (ghalliday) [inline esp/services/WsDeploy/WsDeployPlugin.cpp:47]: is this related?  Something to note in the jira?

File: esp/services/WsDeploy/WsDeployPlugin.cpp [hunk 5]
```diff
@@ -59,7 +59,7 @@ ESP_FACTORY IEspService * esp_service_factory(const char *name, const char* type
 
 ESP_FACTORY IEspRpcBinding * esp_binding_factory(const char *name, const char* type, IPropertyTree *cfg, const char *process)
 {
-   if (strcmp(type, "WsDeploySoapBinding")==0)
+   if (strieq(type, "WsDeploySoapBinding")||strieq(type, "wsdeploy_http"))
```
Comment (ghalliday) [inline esp/services/WsDeploy/WsDeployPlugin.cpp:62]: Does this need a comment why http is also included here?

File: esp/services/ws_access/ws_accessService.hpp [hunk 6]
```diff
@@ -40,6 +40,11 @@ class Cws_accessSoapBindingEx : public Cws_accessSoapBinding
 
     virtual void getNavigationData(IEspContext &context, IPropertyTree & data)
     {
+        if (queryComponentConfig().getPropBool("@api_only"))
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.hpp:43]: These changes aren't obviously related - should they be a separate jira/PR?

File: esp/platform/application_config.cpp [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+static void appendPTreeFromYamlFile(IPropertyTree *tree, const char *file)
+{
+    Owned<IPropertyTree> appendTree = createPTreeFromYAMLFile(file);
+    if (appendTree->hasProp("esp"))
+        mergeConfiguration(*tree, *appendTree->queryPropTree("esp"), nullptr);
+    else
+        mergeConfiguration(*tree, *appendTree, nullptr);
+}
+
+IPropertyTree *loadApplicationConfig(const char *application, const char *defaultYaml, const char* argv[])
```
Comment (ghalliday) [inline esp/platform/application_config.cpp:41]: defaultYaml is unused.

**Review by ghalliday — COMMENTED** [review]
@afishbeck a few initial comments.

Main comment would be that the change could do with some more documentation on how the files are structured, and how they will be configured.  Possible within a esp/READ.md.  It also looks like the configuration files contain quite a bit of duplication which I didn't expect.

File: esp/applications/eclwatch/application.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+   - WsDfu
+   - WsDfuXRef
+   - WsFileIO
+   - WsPackageProcess
+   - FileSpray
+   - ws_machine
+   - ws_elk
+   - ws_access
+   - ws_account
+   - ws_config
```
Comment (richardkchapman) [inline esp/applications/eclwatch/application.yaml:16]: ws_config is not built in containerized mode.

I'm not convinced it should be part of --application=eclwatch at all, but if it is we need to handle having these files version-dependent or copr with references to services that fail to load.

File: esp/platform/application_config.cpp [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        StringBuffer protocolXml("<EspProtocol name='http' type='http_protocol' plugin='esphttp' maxRequestEntityLength='60000000'/>");
+        legacyEsp->addPropTree("EspProtocol", createPTreeFromXMLString(protocolXml));
+    }
+    return useTls;
+}
+
+void addServices(IPropertyTree *legacyEsp, IPropertyTree *appEsp, const char *application, const char *auth, bool tls)
+{
+    Owned<IPropertyTreeIterator> services = appEsp->getElements("application/services");
+    int defPort = tls ? appEsp->getPropInt("@tlsPort", 443) : appEsp->getPropInt("@httpPort", 80); //allows boolean switch between either http or tls config
```
Comment (richardkchapman) [inline esp/platform/application_config.cpp:219]: Not sure 80 is a good default. You can't bind to port 80 unless running as root or special capabilities set.

Comment (richardkchapman) [reply esp/platform/application_config.cpp:219]: Or perhaps httpPort=8010 should be included in the application=eclwatch and application=http_services yaml files,  so that the default port for those applications is what it has always been...

File: esp/applications/eclwatch/eclwatch.yaml (hunk 1 above)
Comment (afishbeck) [reply esp/applications/eclwatch/eclwatch.yaml:13]: This is at least temporarily there for backward compatability.  It creates a node like:  <xslt name="atts">/esp/xslt/atts.xslt</xslt>.  I'm trying not to change the content as part of this pull request, that can come later.

File: esp/applications/eclwatch/eclwatch.yaml [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+      ViewTimeout: 1000
+      LayoutProgram: dot/dot -Tsvg -Gordering=out
+      StyleSheets:
+        xslt:
+        - name: def_file
+          ^: "./smc_xslt/def_file.xslt"
+   FileSpray:
+      QueueLabel: dfuserver_queue
+      MonitorQueueLabel: dfuserver_monitor_queue
+      RootFolder: /c$/thordata
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/applications/eclwatch/eclwatch.yaml:47]: It's probably an old remnant, I'll remove it.

File: esp/applications/eclwatch/esp.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+   httpPort: 80
+   tlsPort: 443
+   enableSEHMapping: true
+   httpConfigAccess: true
+   logLevel: 1
+   maxBacklogQueueSize: 200
+   portalurl: http://hpccsystems.com/download
+
+   tls_config:
+     certificate: /etc/HPCCSystems/certificates/{$name}/server.crt
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/applications/eclwatch/esp.yaml:16]: Definitely agree.  That will be one of my next steps.

File: esp/applications/eclwatch/application.yaml (hunk 8 above)
Comment (afishbeck) [reply esp/applications/eclwatch/application.yaml:16]: I'll remove it.

**Comment by afishbeck** [comment]
@richardkchapman updated and rebased.  Note that the front end eclwatch now defaults to tls port 8843 and http port 8880.  But I think perhaps the helm chart should set up port forwarding of 80 and 443 so that browser support is seamless?

**Comment by afishbeck** [comment]
hpcc_services defaults to 8010 and 18010 to be backward compatible with internal use of services for hthor, thor, roxie, fileservices, etc.

File: esp/applications/eclwatch/eclwatch.yaml (hunk 1 above)
Comment (richardkchapman) [reply esp/applications/eclwatch/eclwatch.yaml:13]: Is this standard yaml? Or an extension in our jptree code?

Comment (afishbeck) [reply esp/applications/eclwatch/eclwatch.yaml:13]: The YAML is valid although it's unlikely anyone else would use this syntax.  For PTREE it would result in an attribute name we couldn't support.  I use that special name to load it into the ptree in a unique way internally, but it gets written back out the same or similar to how you see it here.

So for YAML it's probably very unique, but legal.  I only use it in cases where I need backward compatability.  We can change this when we clean up what the services expect.  In fact this can probably be removed as it's from legacy eclwatch.. but we can focus on that in a follow on PR or point release.

File: esp/applications/eclwatch/ldap_authorization_map.yaml (hunk 2 above)
Comment (afishbeck) [reply esp/applications/eclwatch/ldap_authorization_map.yaml:122]: I don't know.  Once this is final I'll open up a JIRA for reviewing things that might be removed because they are deprecated, etc.

File: esp/applications/hpcc_services/hpcc_services.yaml (hunk 3 above)
Comment (afishbeck) [reply esp/applications/hpcc_services/hpcc_services.yaml]: Purely rebuilding the internal filters for "preflight".  I'm hoping this all goes away or moves for both windows and linux.  Perhaps the preflight pages can be or are removed for containerized?  I imagine we will just use standard cloud tools for monitoring in the future.

File: esp/applications/hpcc_services/hpcc_services.yaml [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+         Store:
+         -  name: HPCCApps
+            description: Generic KeyVal store for HPCC Applications
+            default: true
+   ws_machine:
+      excludePartitions: /dev*,/sys*,/proc*,/run*,/boot
+      monitorDaliFileServer: false
+      UseDefaultSSHUserID: false
+      UseDefaultPIDFileName: true
+      ProcessFilters:
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/applications/hpcc_services/hpcc_services.yaml]: Now called eclservices this is the the internal copy of our webservices for calling via soapcall or ecl plugins, etc.  I made copies of everything because I suspect they will diverge.  Once things are clear we can consider sharing some content between applications.  But I'd rather have the freedom to customize each separately in the beginning.

File: esp/applications/hpcc_services/ldap_authorization_map.yaml [hunk 13]
```diff
@@ -0,0 +1,153 @@
+ldap:
+   root_access:
+      resource: SmcAccess
+      required: Read
+      description: Base access to EclWatch
+   resource_map:
+      WsSMC:
+         Feature:
+         -  path: SmcAccess
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/applications/hpcc_services/ldap_authorization_map.yaml]: Yes the mappings are currently the same.  We can probably share this file between them but wanted to wait to do that once I'm confident they wont diverge.

File: esp/bindings/http/platform/httpbinding.cpp [hunk 14]
```diff
@@ -1539,6 +1539,20 @@ int EspHttpBinding::onGetConfig(IEspContext &context, CHttpRequest* request, CHt
     ISecUser* user = context.queryUser();
     if (m_viewConfig || (user && (user->getStatus()==SecUserStatus_Inhouse)))
     {
+        if (getESPContainer() && getESPContainer()->queryApplicationConfig())
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/bindings/http/platform/httpbinding.cpp:1542]: This is primarily used for backward compatibility so that when a user asks to see the config I can give them the correct information.  ESP used to just return the xml file, but the information is generated now so I need to return it.  It will definitely help when testing the new application mode to see how the config differs from what it would have been via legacy configmgr.

File: esp/services/WsDeploy/WsDeployPlugin.cpp (hunk 4 above)
Comment (afishbeck) [reply esp/services/WsDeploy/WsDeployPlugin.cpp:47]: Could have done without changing these calls.

File: esp/services/ws_access/ws_accessService.hpp (hunk 6 above)
Comment (afishbeck) [reply esp/services/ws_access/ws_accessService.hpp:43]: This is initially what makes eclservices different from eclwatch.  This turns off the UI and restores the default web service forms.
I expect the two to diverge further over time.
I could have separated the creation of eclservices from the ability to run as an application but thought it would be a good exercise in making sure the feature met the needs of dealing with different flavors of application.

File: esp/services/WsDeploy/WsDeployPlugin.cpp (hunk 5 above)
Comment (afishbeck) [reply esp/services/WsDeploy/WsDeployPlugin.cpp:62]: The existing form for the name was inconsistent and therefor hard to automatically generate.  _http also more accurately reflects that these bindings cover all HTTP based protocol bindings not just SOAP.  I'll add comments.

File: esp/platform/espp.cpp [hunk 15]
```diff
@@ -311,10 +315,29 @@ static Owned<IPropertyTree> espConfig;
...
 static void usage()
 {
-    puts("ESP - Enterprise Service Platform server. (C) 2001-2011, HPCC Systems®.");
+    puts("ESP - Enterprise Service Platform server. (C) 2001-2020, HPCC Systems®.");
     puts("Usage:");
+    puts("To start ESP in application mode");
+    puts("  esp --aplication=<appname> [options]");
+    puts("Example Application names:");
+    puts("  eclwatch - front end ECL user interface and services");
+    puts("  hpcc_services - back end ECL system services");
```
Comment (richardkchapman) [inline esp/platform/espp.cpp:324]: Should change this to reflect new name

**Comment by afishbeck** [comment]
Updated.  Changed the name of the wsecl application to "eclqueries", and cleaned up usage text.
